File: backend/models/options.py
from pydantic import BaseModel, Field, validator, ConfigDict
from datetime import datetime
from typing import Optional, Literal

class OptionBase(BaseModel):
    symbol: str
    strategy_type: str
    trade_direction: Literal['Bullish', 'Bearish', 'Neutral']
    number_of_contracts: int = Field(..., gt=0)
    option_type: Literal['Call', 'Put']
    strike_price: float = Field(..., gt=0)
    expiration_date: datetime
    entry_price: float = Field(..., gt=0)
    exit_price: Optional[float] = None
    total_premium: float = Field(..., ge=0)
    commissions: float = Field(..., ge=0)
    implied_volatility: float = Field(..., ge=0, le=1000)  # IV as percentage (e.g., 25.5 for 25.5%)
    entry_date: datetime
    exit_date: Optional[datetime] = None
    status: Optional[Literal['open', 'closed']] = 'open'

class OptionCreate(OptionBase):
    pass

class OptionUpdate(BaseModel):
    exit_price: Optional[float] = None
    exit_date: Optional[datetime] = None
    status: Optional[Literal['open', 'closed']] = None

class OptionInDB(OptionBase):
    id: int
    user_id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: lib/types/options.ts
export type OptionTradeStatus = 'open' | 'closed';
export type OptionType = 'Call' | 'Put';
export type TradeDirection = 'Bullish' | 'Bearish' | 'Neutral';

export interface OptionTrade {
  id: number;
  symbol: string;
  strategyType: string;
  tradeDirection: TradeDirection;
  numberOfContracts: number;
  optionType: OptionType;
  strikePrice: string; // DECIMAL as string
  expirationDate: string; // ISO
  entryPrice: string; // DECIMAL as string
  exitPrice?: string | null;
  totalPremium: string; // DECIMAL as string
  commissions: string; // DECIMAL as string
  impliedVolatility: string; // DECIMAL as string
  entryDate: string; // ISO
  exitDate?: string | null; // ISO
  status: OptionTradeStatus;
  initialTarget?: string | null;
  profitTarget?: string | null;
  tradeRatings?: number | null; // 1-5 stars
  createdAt: string;
  updatedAt: string;
  reviewed: boolean;
  mistakes?: string | null;
}

export interface CreateOptionRequest {
  symbol: string;
  strategyType: string;
  tradeDirection: TradeDirection;
  numberOfContracts: number;
  optionType: OptionType;
  strikePrice: number;
  expirationDate: string; // ISO
  entryPrice: number;
  totalPremium: number;
  commissions?: number;
  impliedVolatility: number;
  entryDate: string; // ISO
  initialTarget?: number;
  profitTarget?: number;
  tradeRatings?: number; // 1-5 stars
  reviewed?: boolean;
  mistakes?: string;
}

export interface UpdateOptionRequest {
  symbol?: string;
  strategyType?: string;
  tradeDirection?: TradeDirection;
  numberOfContracts?: number;
  optionType?: OptionType;
  strikePrice?: number;
  expirationDate?: string;
  entryPrice?: number;
  exitPrice?: number | null;
  totalPremium?: number;
  commissions?: number;
  impliedVolatility?: number;
  entryDate?: string;
  exitDate?: string | null;
  status?: OptionTradeStatus;
  initialTarget?: number | null;
  profitTarget?: number | null;
  tradeRatings?: number | null; // 1-5 stars
  reviewed?: boolean;
  mistakes?: string | null;
}


